### PR TITLE
PIL-977 -  Agent Authorisation for Phase 2

### PR DESCRIPTION
**Pull request by TevynAllen**

### CODE REVIEW

File: app/controllers/AgentController.scala [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  def onSubmitClientPillarId: Action[AnyContent] = (featureAction.asaAccessAction andThen agentIdentify() andThen getData andThen requireData).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(clientPillarIdView(formWithErrors))),
+          value => {
+            val result = for {
+              updatedAnswers     <- Future.fromTry(request.userAnswers.set(AgentClientPillar2ReferencePage, value))
+              _                  <- sessionRepository.set(updatedAnswers)
```
Comment (alikouhestani) [inline app/controllers/AgentController.scala:79]: Isn't it better to just do 
```suggestion
              subscriptionData   <- subscriptionService.readSubscription(value)
              answersWithOrgName <- Future.fromTry(updatedAnswers.set(AgentClientOrganisationNamePage, subscriptionData.upeDetails.organisationName))
                            _                  <- sessionRepository.set(answersWithOrgName)
```
? Not sure why you are calling set twice

File: app/controllers/auth/AuthController.scala [hunk 2]
```diff
@@ -17,37 +17,67 @@
...
-  def signOut(): Action[AnyContent] = identify.async { implicit request =>
-    sessionRepository
-      .clear(request.userId)
+  def signOut(): Action[AnyContent] = Action.async { implicit request =>
+    authorised(AuthProviders(GovernmentGateway))
+      .retrieve(Retrievals.internalId) {
+        case Some(internalId) =>
+          sessionRepository.clear(internalId)
+        case _ =>
+          logger.warn(s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Unable to retrieve internal id or affinity group")
```
Comment (alikouhestani) [inline app/controllers/auth/AuthController.scala:50]: You don't need to log session ID, this is logged via MDC and appears on a separate searchable field

File: app/controllers/auth/AuthController.scala [hunk 3]
```diff
@@ -17,37 +17,67 @@
...
-  def signOutNoSurvey(): Action[AnyContent] = identify.async { implicit request =>
-    sessionRepository
-      .clear(request.userId)
+  def signOutNoSurvey(): Action[AnyContent] = Action.async { implicit request =>
+    authorised(AuthProviders(GovernmentGateway))
+      .retrieve(Retrievals.internalId) {
+        case Some(internalId) =>
+          sessionRepository.clear(internalId)
+        case _ =>
+          logger.warn(s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Unable to retrieve internal id or affinity group")
```
Comment (alikouhestani) [inline app/controllers/auth/AuthController.scala:70]: Same here

File: app/controllers/AgentController.scala (hunk 1 above)
Comment (TevynAllen) [reply app/controllers/AgentController.scala:79]: Not sure what you mean by calling set twice?

1. The first set is saving the pillar2 id the agent entered in the form
2. We use that ID to call read subscription to get the organisation name and also verify that there is an record for that ID (If this fails the agent is redirected to a KB page where they can go back and re-enter a new ID or alter the one they entered for instance they had a typo) 
3. If successful we want to save the organisation name and redirect them to the next page in the journey

File: app/controllers/auth/AuthController.scala (hunk 2 above)
Comment (TevynAllen) [reply app/controllers/auth/AuthController.scala]: Correct 👍 I copy and pasted it from the action

**Review by alikouhestani — DISMISSED** [review]

**Review by jassalrichy — APPROVED** [review]

**Pull request merged**